### ISSUE TITLE
Question: slots with data

### ISSUE THREAD

**Issue by Marvin-Brouwer** [BuilderIO/mitosis#1333]
I need a slot in a list in which I can pass around the data of it's parent.
So for example, with a component like this:
```jsx
import { Slot, Layout, For } from "@builder.io/mitosis";

export default function MyComponent(props) {
  const example = [ 

    "asdf",
    "dsfdsdf"
  ]

  return (
    <div>
      
      <For each={example} >
        {
          (item) => (
            <Slot name={props.slotPage} item={item}></Slot>
          )
        }
      </For>
    </div>
  );
}
```
I'd like this ouput in Vue
```vue
<template>
  <div>
    <template :key="index" v-for="(item, index) in example">
      <slot name="page" :item={item}></slot>
    </template>
  </div>
</template>

<script>
import { defineComponent } from "vue";

export default defineComponent({
  name: "my-component",
});
</script>
```
In vue you can do: `<slot name="page" :item={item}></slot>`, however I can't seem to figure out how to make mitosis do this.
The documentation doesn't state this as a gotcha, but it also doesn't state how to do this.

If I need to make a feature request for this, please just tell me and close the ticket.

Thanks,

----
https://mitosis.builder.io/?outputTab=G4VwpkA%3D&code=JYWwDg9gTgLgBAbzgZQDYRgGjgGQIYCeEArlnAGLRwC%2BcAZlBCHAEQACARscKgCYCmUAHTAIAehDAYEAM7AZLANwAoZfwAekWHAF08xVPDrEAdgGMYok3ACyBAMJNIJ%2FiZgAKMIzAyAlImU4ODMIExl4DTxwVH44AF44AG04VSCgljwZXjoWTEC0ll4ZOiLslnyAXVS4KH4YYihrd3yggB5eYAA3AD4WtL62yig4fjwzAAs4hEjo%2FlpetMWApaX3KX4QfzjuuGaVldawbta0DDgTKP4prwgfIRl0GAAFPABzObh1kCmv6mOxU4wf5HAZLXygoLUUGtMRDBZpGEdHr5XwqahAA%3D%3D%3D&inputTab=FYZwHkA%3D

**Comment by samijaber** [comment]
This is a bug. Here we generate these slots: https://github.com/BuilderIO/mitosis/blob/e862056eca675887bd975bf5696c61666d0f3532/packages/core/src/generators/vue/blocks.ts#L297-L305

We are missing the slots' properties and bindings, which are generated by this function:

https://github.com/BuilderIO/mitosis/blob/e862056eca675887bd975bf5696c61666d0f3532/packages/core/src/generators/vue/blocks.ts#L378

I believe you should be able to reuse that function. If not, feel free to modify it as needed

**Comment by Marvin-Brouwer** [comment]
I'm not sure how to "reuse that function", do you mean when creating a bugfix?

**Comment by samijaber** [comment]
Yes, I am describing the process of contributing a fix to Mitosis itself 😄

**Comment by Marvin-Brouwer** [comment]
AH, check.
I might take some time somewhere next week to do so.

I see react and svelte also aren't binding this, is this something you'd want?
And for example for react, would this mean it would output like `<p>{slotExample(value)}<p/>`
or would we maybe want to include a <Slot> component for react?
And do we know of more components that might need an update?

**Comment by Marvin-Brouwer** [comment]
@samijaber I pulled down the repo and I immediately have failing tests:
![image](https://github.com/BuilderIO/mitosis/assets/5499778/488e8ea0-615d-427d-8372-8cd93397a34a)
Is there anything I need to configure?

**Comment by samijaber** [comment]
There might be a problem with the configuration of the repo. Do the tests pass if you run `yarn ci:build` in the root?

**Comment by Marvin-Brouwer** [comment]
Yes that seems to work out fine.
So running `yarn test` in the `~/packages/core` folder doesn't work as expected.

Additionally, I now have 22 modified files after running `yarn ci:build` in the root.
Are you on unix or mac by any chance?
All of the changes appear to be whitespace.

**Comment by Marvin-Brouwer** [comment]
@samijaber I'm having a hard time understanding this part:
![image](https://github.com/BuilderIO/mitosis/assets/5499778/17a82a5b-76fd-4951-8654-a9d13ddc6c27)

It looks like that in the case of a nameless slot, when there is anything in `json.bindings` the vue template mechanism should be used.
However, these don't seem to be specified in the snapshot tests.
How is this supposed to behave? And what about when someone would add a `<Slot key="whatever" ...` on their component? Should it also handle that key property?

**Comment by Marvin-Brouwer** [comment]
@samijaber I solved it like this for now:
![image](https://github.com/BuilderIO/mitosis/assets/5499778/72979a25-e171-41ba-99e1-6ca3dbd142cf)
Please let me know if this is correct or not.
I'll add some documentation about this too.
However, I am kind of concerned this behavior doesn't really exist outside of Vue.

**Comment by Marvin-Brouwer** [comment]
Hi, 

@samijaber I've figured some things out and made some progress but I'm kind of stuck right now.
When generating the snapshots somehow there's a type definition generated:
![image](https://github.com/BuilderIO/mitosis/assets/5499778/ad1108cf-5e77-4b5a-8ffa-030814d62a7b)

I can't find where this happens.
Could you please point me to the block of code that is responsible for generating these propType definitions?

Thanks

**Comment by samijaber** [comment]
These definitions are stored in the `types` key, and the name of the type (`Props` in this case) is stored in `propsTypesRef`:

https://github.com/BuilderIO/mitosis/blob/e862056eca675887bd975bf5696c61666d0f3532/packages/core/src/types/mitosis-component.ts#L148-L149

which is then consumed by each generator, like here:

https://github.com/BuilderIO/mitosis/blob/e862056eca675887bd975bf5696c61666d0f3532/packages/core/src/generators/react/generator.ts#L561

**Comment by samijaber** [comment]
If you are concerned about the `slot` prefix added to slots in React, that is being fixed right now in #1334

**Comment by Marvin-Brouwer** [comment]
No I'm not, I converted react slots from a field to function so data can passed just like in vue, svelte, etc.

However, the type still generates `JSX.Element` instead of `() => JSX.Element` or `(slotProps) => JSX.Element`
So, I'd like to take a stab at rectifying that.

**Comment by samijaber** [comment]
I just looked at the snapshot you screenshotted:

https://github.com/BuilderIO/mitosis/blob/e862056eca675887bd975bf5696c61666d0f3532/packages/core/src/__tests__/data/blocks/content-slot-html.raw.tsx#L2-L8

It is actually an error to import the Mitosis `JSX.Element` here. For now, the best we can do is make this `slotTesting: any`, because Mitosis does not properly handle generating element types across all frameworks.

So it is safe to ignore this type or change it to `any` in the `.raw.tsx` file if you prefer.

**Comment by Marvin-Brouwer** [comment]
Ah, check, thanks.
I'll continue with solid then.

I may have some backwards compatibility questions when I'm done.
But I'll first see how far I can get this to work across all output types.

**Comment by samijaber** [comment]
@Marvin-Brouwer I just re-read your implementation suggestion and realized it could cause breaking issues for other folks.

Would you mind firing up a draft PR with what you have so far, so I can see what you mean exactly and align on the implementation? This will prevent you from wasting time in a direction we don't want Mitosis to go into.

**Comment by Marvin-Brouwer** [comment]
Sure, on it!

**Comment by Marvin-Brouwer** [comment]
Here you go: https://github.com/BuilderIO/mitosis/pull/1336
I included some comments about the backwards compatibility thing I'd like to address if you decide to go forward with this.

**Comment by Marvin-Brouwer** [comment]
@samijaber 
Any news?

**Comment by Marvin-Brouwer** [comment]
@samijaber is there anything I can do to help move this forward?

**Comment by Murderlon** [comment]
Another friendly ping. This is practically blocking me from adopting Mitosis right now.
